### PR TITLE
New `RingPedersen` API

### LINKED ISSUE #43: Add docs and unit tests to ZKP setup
The `ZkSetupParameters` struct and corresponding methods are missing a few code quality items.

- [ ] Add struct documentation. Explain the role of the parameters.
- [ ] Improve naming of fields. Replace single-letter names (`N`, `s`, `t`) with more meaningful names and add the single-letter names to the field docs.
- [ ] Improve types of fields. `N` should be a `PaillerEncryptionKey`; the type of `s` and `t` is blocked by #40  (can leave as-is for now, since 40 includes fixing it.).
- [ ] Add unit test. If a value is created from `gen_from_primes`, it should `verify`.
- [ ] Revisit `gen_from_primes` API. The parameter `N` could be passed by value (so the caller can decide whether to clone it or not), or we could skip passing `N` because it's formed from `p` and `q`. Or, we might form it from the `PaillierDecryptionKey` instead of `(N, p, q)`, which already has the totient and `N`.

### LINKED ISSUE #128: Make Ring-Pedersen commitment abstraction
Many proofs involve Ring-Pedersen commitments. Right now, these are computed manually throughout the protocol, using the `ZkSetupParameters` to hold the Ring-Pedersen parameters (as well as a proof that they are valid). Standard commitments and operations on commitments are computed manually in many places, which makes it (a) harder to recognize them as commitments and (b) easier to mis-compute at some point.

To complete this issue:
- [ ] make a Ring-Pedersen module with an appropriate API, documentation, and tests
- [ ] integrate the module into the protocol (this might be better spun out into a separate issue if the PR starts to get huge)

## Proposed solution
@amaloz wrote the following API for Ring-Pedersen commitments. This would replace manual / inline computations across all the protocols, and would replace any tuples `(N, s, t)`, such as the one in `ZkSetupParameters`, with a `RingPedersen` object.

```rust
pub(crate) struct RingPedersen {
    /// The RSA modulus.
    pub(crate) N: BigNumber,
    /// Ring-Petersen parameter `s`.
    pub(crate) s: BigNumber,
    /// Ring-Petersen parameter `t`.
    pub(crate) t: BigNumber,
}

#[derive(PartialEq)]
pub(crate) struct Commitment(BigNumber);

impl RingPedersen {
    /// Extracts a [`RingPedersen`] object from a [`PaillierDecryptionKey`].
    pub(crate) fn derive(sk: &PaillierDecryptionKey) -> Result<Self> {}

    /// Produces a Pedersen commitment $`s^a t^b`$ commiting to the values `a` and `b`.
    pub(crate) fn commit(&self, a: &BigNumber, b: &BigNumber) -> Commitment {}

    /// Produces a Pedersen commitment $`s^a t^b`$ for values $`a \in 2^{\text{range}_0}`$
    /// and $`b \in 2^{\text{range}_1}`$.
    pub(crate) fn commit_from_ranges(
        &self,
        range0: usize,
        range1: usize,
        rng: &mut (impl RngCore + CryptoRng),
    ) -> Result<(Commitment, BigNumber, BigNumber)> {
    }

    /// Combines two commitments with exponent `e` by computing $`\text{com}_0 \text{com}_1^{e}`$.
    pub(crate) fn combine(
        &self,
        com0: &Commitment,
        com1: &Commitment,
        e: &BigNumber,
    ) -> Result<Commitment> {
    }
}
```

### PR DESCRIPTION
**Pull request by amaloz**
This PR introduces a new `RingPedersen` object, which encapsulates the ring-Pedersen commitment scheme used throughout various ZKPs. It allows for creating new commitments, reconstructing commitments, and combining commitments. Using this API greatly cleans up code in the various ZKPs that use ring-Pedersen commitments.

Unfortunately, one of the ZKPs (`pifac`) uses the commitments in non-standard ways: instead of computing a commitment as `s^a t^r` it computes `c^a t^r`, where `c` is an existing commitment. This sadly requires a whole set of API calls to support. These are named `commit_with_commitment` and `reconstruct_with_commitment`. I don't yet have a strong opinion on whether we should retain these are not, but they are there as a demonstration of what the full API could look like.

Closes #128. Closes #43 by fact of removing that component.

### CODE REVIEW

File: src/ring_pedersen.rs
Comment (amaloz) [inline src/ring_pedersen.rs:1]: Tests to be added once overall design agreed upon.

File: src/ring_pedersen.rs
Comment (marsella) [inline src/ring_pedersen.rs:74]: suggestion: When we use this, we just need a reference -> reference transformation.  

I was going to suggest `AsRef` instead of `From`, but actually, since this kind of breaks our abstractions, I might prefer a custom method like `as_masked(&self) -> &MaskedRandomness` that has specific guidance in the docs for how you should use it e.g. 

> Use for the rare case where you need to know / commit to the nonce before you generate a [`Commitment`]. The nonce won't really be "masked" with anything, but this will let you call `reconstruct_with_commitment` on the known nonce.


Edit: Maybe the types don't work with the ref->ref situation, but I'd prefer a custom method over an undocumented trait impl anyway.

File: src/ring_pedersen.rs
Comment (marsella) [inline src/ring_pedersen.rs:80]: suggestion: I think I saw in the usage that you downgrade masked -> commitment randomness and then immediately mask it again. Can we just make a `mask` or `remask` method on `MaskedRandomness` instead?

File: src/ring_pedersen.rs
Comment (marsella) [inline src/ring_pedersen.rs:97]: suggestion: We often call methods like this `from_parts`, but I'm not sure if that's an internal convention or more general.

**Review by marsella — COMMENTED** [review]
This looks great! I like the names you've chosen throughout, it makes the proofs much easier to read.

I have two API change requests and I also dove in a bit to specific code changes. There are more details inline, but the overview is:
1. I want more explicit, documented methods for the rare cases where we manually convert between Masked and Commitment randomness
2. I think we should consider pulling `ZKSetupParameters` into this module, so there's no chance to separate the proof from the parameters.

I think the natural outcome of an API like this is slightly more difficult auditability -- particularly for `PiFacProof`, the auditor will have to read this module carefully to make sure we're sampling from all the right distributions for all the little values. However, the improved readability and encapsulation seems worth it to me. I'll do a second review on this to make sure we've got it right (didn't see any issues on this pass but still), and I think we'll be good to go.

**Comment by marsella** [comment]
Another thought I had while reading through this is that our crypto APIs have expanded enough that we might benefit from reshuffling our structure a bit and making a `core` or `crypto` module to contain this, the Paillier module, the ZStarN module, and maybe all the sampling methods that are currently stuffed into `utils`. I'll make a separate issue for that.

File: src/ring_pedersen.rs
Comment (amaloz) [reply src/ring_pedersen.rs:115]: Can do! My take in the past has been if the error is a mistake in using the API (namely, something that should never happen), it's a panic (since no correct program should ever hit that condition), and if it's a recoverable error in any way, it's an error. I'm fine with returning `InternalInvariantFaield` though, just wanted to explain the reasoning here to see if it's something we'd want to adopt at some point :-).

File: src/ring_pedersen.rs
Comment (amaloz) [reply src/ring_pedersen.rs:116]: Oh this is much better!

File: src/zkp/piprm.rs
Comment (marsella) [reply src/zkp/piprm.rs:136]: That's fair. I'll add it as a bullet point to #51, we can fix it there.

File: src/ring_pedersen.rs
Comment (marsella) [reply src/ring_pedersen.rs:115]: Since we're running this as a library in production, we need to have all our errors be recoverable, even when it's a bug on our end. @gatoWololo  is working on an updated scheme for passing a limited set of actionable errors to the caller and logging all the useful context for the developer. But in the end, if we panic here, we might cause panics up some massive chain of production servers and we can't really do that responsibly.

Comment (amaloz) [reply src/ring_pedersen.rs:115]: Makes sense!

File: src/zkp/setup.rs
Comment (amaloz) [reply src/zkp/setup.rs]: Yes, this is way cleaner and removes all the `setup.rs` code!

File: src/ring_pedersen.rs
Comment (amaloz) [reply src/ring_pedersen.rs]: Good idea. I was able to fix this using `transmute` to retain the references, but that requires unsafe code. If we're okay using the `bytemuck` crate we can do this in a safe manner: https://docs.rs/bytemuck/latest/bytemuck/derive.TransparentWrapper.html

Comment (marsella) [reply src/ring_pedersen.rs]: `bytemuck` seems fine to me. @gatoWololo do you have any concerns?

File: src/ring_pedersen.rs
Comment (marsella) [inline src/ring_pedersen.rs:203]: nits:
- `commit` doesn't explain what it's computing :) Please add the formula up there.
- newline between first and second sentence.

File: src/ring_pedersen.rs
Comment (marsella) [inline src/ring_pedersen.rs:189]: nit: 
- Update the wording to be more explicit about what is and isn't enforced. E.g. the expected usage is that `value` is in range, but we don't check (should we check?).
- Per other comment, add the formula for commitments.
- Adjust phrasing to have a one-liner method description, then newline, then all these other details.
```
/// Produces a commitment to `value`.
///
/// blah blah in range
```

File: src/ring_pedersen.rs
Comment (marsella) [inline src/ring_pedersen.rs:262]: nit: this is a personal pet peeve, but I prefer to name Rust tests without the word `test` in them (since they're run with `cargo test` and are labeled with `#[test]`, we can basically always tell that they're tests). 

I like names that describe what's being checked, like
- `verified_ring_pedersen_generation_works`
- `parameters_must_match_proof_to_verify`
etc.

File: src/ring_pedersen.rs
Comment (marsella) [inline src/ring_pedersen.rs:1]: todo: Add module-level docs (at least a one-line description, and optionally more context about commitment schemes and expected usage of this module) and a copyright notice.

File: src/auxinfo/participant.rs
Comment (marsella) [inline src/auxinfo/participant.rs:512]: This line has annoyed me so many times. I'm very grateful that it's gone.

**Review by marsella — CHANGES_REQUESTED** [review]
I have some small changes, but they're almost all documentation and naming.

**Comment by marsella** [comment]
I think this also closes #43 by removing the `ZkSetupParameters` struct and having reasonable documentation on `VerifiedRingPedersen`.

File: src/ring_pedersen.rs
Comment (amaloz) [reply src/ring_pedersen.rs:189]: > should we check?

For robustness it's probably a good idea that we check I think. It might hurt performance though, so there's a tradeoff. Any thoughts?

Comment (amaloz) [reply src/ring_pedersen.rs:189]: I added a test for this, expecting the commitment to not reconstruct correctly, but it does. Which actually makes sense, there's nothing in the commitment process that restricts `value`'s range.

This suggests to me it makes even more sense to add a check for misuse resistance.

File: src/ring_pedersen.rs
Comment (amaloz) [reply src/ring_pedersen.rs:1]: What copyright notice should I use? Presumably it shouldn't be the facebook one, should it be:
```
// Copyright (c) Bolt Labs, Inc.
//
// This source code is licensed under both the MIT license found in the
// LICENSE-MIT file in the root directory of this source tree and the Apache
// License, Version 2.0 found in the LICENSE-APACHE file in the root directory
// of this source tree.
```

File: src/ring_pedersen.rs
Comment (amaloz) [reply src/ring_pedersen.rs:79]: People tend to get scared by `unsafe` code (probably rightfully so) so if we can avoid it all the better. `bytemuck` is a fairly standard library and would allow us to avoid this, so that might be the best approach.

**Comment by amaloz** [comment]
> I think this also closes #43 by removing the `ZkSetupParameters` struct and having reasonable documentation on `VerifiedRingPedersen`.

The one bit not handles is the use of `ZStar`. I can add that in to this PR or do that in another one (I'd prefer another one, since this one is already quite large).

File: src/ring_pedersen.rs
Comment (amaloz) [reply src/ring_pedersen.rs]: Hmm, this is trickier than I thought. `reconstruct` also takes a `value`, but we can't restrict that `value`'s range because it doesn't take `range` as an argument. So is it worth restricting `value`'s range in `commit`, particularly if there's no a priori reason to have that restriction (that is, `commit` works fine over any `value`, not only `value`s in the provided range)?

Comment (amaloz) [reply src/ring_pedersen.rs]: I take it all back (I think): The restriction on `value` causes the proofs to fail :-(. The reason is in the `pifac` step of constructing `R = s^{N_0} t^\sigma`, `N_0` is outside the expected range... To me the cleanest approach would be to remove the restriction on value, but I'm open to other ideas.

Comment (marsella) [reply src/ring_pedersen.rs]: Hm. Some tension between general-purpose Ring-Pedersen and application-specific commitments we're dealing with here. Let's remove the restriction on `value`. The restrictions vary by use case (even within our application) so I think it's reasonable to push the burden of maintaining that correctness to the proofs that call this API. It's 3 proofs with restrictions and 1 without, which makes it a bit annoying to generalize the API to the one. But we both already audited all the places where we call this to make sure it's correct.

File: src/ring_pedersen.rs
Comment (marsella) [reply src/ring_pedersen.rs]: Yeah, that looks good. Add 2023 after the `(c)`, please.

**Comment by marsella** [comment]
> > I think this also closes #43 by removing the `ZkSetupParameters` struct and having reasonable documentation on `VerifiedRingPedersen`.
> 
> The one bit not handles is the use of `ZStar`. I can add that in to this PR or do that in another one (I'd prefer another one, since this one is already quite large).

That should be handled by #132, which @Ra1issa will be taking on after this is merged.

**Comment by amaloz** [comment]
@marsella Should be ready for another look!

File: src/ring_pedersen.rs
Comment (marsella) [inline src/ring_pedersen.rs:10]: nit: Please make the first sentence short so it's easy to scan in the built docs. Like
```rust
//! A verifiable ring-Pedersen commitment scheme.
//!
//! This is a ring variant of the Pedersen commitment scheme, as specified ...

File: src/ring_pedersen.rs
Comment (marsella) [inline src/ring_pedersen.rs:34]: nit: Petersen -> Pe**d**ersen

File: src/zkp/pimod.rs
Comment (marsella) [inline src/zkp/pimod.rs:784]: nit: the `encryption_key()` call could go away now, since we can access the modulus directly from decryption key. Feel free to postpone this to #50, though.

**Review by marsella — APPROVED** [review]
Looks great, thank you!

File: src/zkp/pimod.rs
Comment (amaloz) [reply src/zkp/pimod.rs:784]: Good call. I'll leave it for #50 though.

**Pull request merged**
